File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/312-burst-balloons-approach3.go
package main

import "fmt"

func maxCoins(nums []int) int {
	padded := make([]int, len(nums)+2)
	padded[0] = 1
	padded[len(padded)-1] = 1
	for i, v := range nums {
		padded[i+1] = v
	}
	n := len(padded)
	dp := make([][]int, n)
	for i := range dp {
		dp[i] = make([]int, n)
	}
	for length := 2; length <= n; length++ {
		for i := 0; i <= n-length; i++ {
			j := i + length - 1
			for k := i + 1; k < j; k++ {
				coins := padded[i]*padded[k]*padded[j] + dp[i][k] + dp[k][j]
				if coins > dp[i][j] {
					dp[i][j] = coins
				}
			}
		}
	}
	return dp[0][n-1]
}

func main() {
	assert := func(cond bool) {
		if !cond {
			panic("assertion failed")
		}
	}
	assert(maxCoins([]int{3, 1, 5, 8}) == 167)
	assert(maxCoins([]int{1, 5}) == 10)
	assert(maxCoins([]int{5}) == 5)
	assert(maxCoins([]int{3, 3}) == 12)
	assert(maxCoins([]int{1, 1, 1}) == 3)
	fmt.Println("all tests pass")
}

File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/312-burst-balloons-approach3.py
def max_coins(nums: list[int]) -> int:
    nums = [1] + nums + [1]
    n = len(nums)
    dp = [[0] * n for _ in range(n)]
    for length in range(2, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            for k in range(i + 1, j):
                coins = nums[i] * nums[k] * nums[j] + dp[i][k] + dp[k][j]
                if coins > dp[i][j]:
                    dp[i][j] = coins
    return dp[0][n - 1]

assert max_coins([3, 1, 5, 8]) == 167
assert max_coins([1, 5]) == 10
assert max_coins([5]) == 5
assert max_coins([3, 3]) == 12
assert max_coins([1, 1, 1]) == 3
print("all tests pass")
